### ISSUE TITLE
Sequence of execution

### ISSUE THREAD

**Issue by wfgomes** [snelg/cakephp-3-oracle#3]
Hi, I am trying to do an UPDATE and I got this error

Warning (2): oci_fetch_assoc(): ORA-24374: define not done before fetch or execute and fetch [ROOT/vendor/yajra/laravel-pdo-via-oci8/src/yajra/Pdo/Oci8/Statement.php, line 231]
Warning (2): oci_error() expects parameter 1 to be resource, null given [ROOT/vendor/yajra/laravel-pdo-via-oci8/src/yajra/Pdo/Oci8/Statement.php, line 610]
Notice (8): Undefined variable: _SESSION [CORE/src/Network/Session.php, line 440]

**Comment by heat** [comment]
try search  here https://github.com/yajra/pdo-via-oci8

**Comment by mklappen** [comment]
wfgomes, were you ever able to resolve your error? I receive this error when trying to insert, but not clear to me from looking at the pdo-oci8 statement.php

**Comment by kasmonaut** [comment]
Unfortunately no, I think the error comes from oracle driver.  If you turn off debug mode in CAKE you should be able to update and insert record, but that's not a solution. In my view cake debug kit interferes with current state of driver

**Comment by mklappen** [comment]
Thanks... I did try turning off debug since it comes back as "warning" but for me its still not inserting. Think there's something else going wrong. index actions (that would use select statements) are working just fine.

**Comment by kasmonaut** [comment]
Did you create on insert trigger to populate id field in you oracle table?

**Comment by mklappen** [comment]
Hi, thanks for reply. Yes I setup both a sequence on table and trigger on the id/PK for the table. 
I am fairly certain that the issue may be with the "created" and "modified" fields from cakephp being passed to DB as yyyy-mm-dd HH:mm:ss format when the the oracle database we are using is setup for dd-MMM-yy HH:mm:ss format for some reason. Cant change the DB so will have to change format being sent from cake...haven't had chance to dig into code yet since my last post but hopefully will this weekend. thanks again.

**Comment by kasmonaut** [comment]
Inside a driver in the connection there is command to to set session.
ALTER SESSION SET NLS_DATE_FORMAT='YYYY-MM-DD HH24:MI:SS'
NLS_TIMESTAMP_FORMAT='YYYY-MM-DD HH24:MI:SS'
NLS_TIMESTAMP_TZ_FORMAT='YYYY-MM-DD
HH24:MI:SS'";

make sure it's being executed during connection. This takes care of date
formats.

**Comment by mklappen** [comment]
Thanks so much... that helps steer me in another direction. Also ruled this out since the "modified" column is updating fine when an update query is run. It's only the insert query that's causing me issues. going to keep debugging trying to find what it could be... trigger seems to be working fine on manual inserts run form sql developer (taking the SQL from the debugkit sql log.)
thanks again.

**Comment by mklappen** [comment]
After 6+ hrs of troubleshooting and I'm still no closer to getting a cakephp add function to insert into oracle. I have scaled things back to the most basic and still no luck. The save function in the controller Add function is returning false every time yet an Edit and Index actions work just fine. 

The SQL statement returned from the debug kit logs works just fine in SQL developer so my sequence/trigger are working as well.
<code>
INSERT INTO deployments (
  application, build_no, created, modified
) 
VALUES 
  (
    'qqq', 'qqq', '2016-01-17 00:14:16', 
    '2016-01-17 00:14:16'
  )</code>

running debug on the $query and $queryString in plugin's Oracle.php driver returns what appears to be valid SQL. 
<code>$queryString = 'INSERT INTO deployments (application, build_no, created, modified) VALUES (:c0, :c1, :c2, :c3)'</code>
<code>$query = 'sql' => 'INSERT INTO deployments (application, build_no, created, modified) VALUES (:c0, :c1, :c2, :c3)',
    'params' => [
        ':c0' => [
            'value' => 'qqq',
            'type' => 'string',
            'placeholder' => 'c0'
        ],
        ':c1' => [
            'value' => 'qqq',
            'type' => 'string',
            'placeholder' => 'c1'
        ],
        ':c2' => [
            'value' => object(Cake\I18n\Time){
                'time' => '2016-01-17T00:25:14+0000',
                'timezone' => 'UTC',
                'fixedNowTime' => false
            },
            'type' => 'timestamp',
            'placeholder' => 'c2'
... etc... } </code>

Any thoughts, tips or insights is greatly appreciated as I have exhausted my oracle knowledge at this point.

**Comment by mklappen** [comment]
The issue appears to be with the _newId function call in cakePHP's table.php. Something is causing memory to exceed limits (regardless of how big I set PHP's memory_limit too). it's happening in the

<code>$type = Type::build($typeName);</code>
<code>debug($primary) outputs [0] => 'id'</code>
<code>debug($typeName) outputs 'integer'</code>
so not sure what is wrong, the datatype in Oracle is set to number (size = 10) but when I try to debug $type script blows up with Fatal Error. exceeded memory limits...
<code>
    protected function _newId($primary)<br />
    {
        debug($primary);
        if (!$primary || count((array)$primary) > 1) {
            return null;
        }
        $typeName = $this->schema()->columnType($primary[0]);
        $type = Type::build($typeName);
    debug($type);
        return $type->newId();
    }
</code>

**Comment by kasmonaut** [comment]
Sorry, I do not have any more incites on this issue. What I tried in parallel is to use actual PDO  driver from Oracle, not the one in this project based on OCI.  The PDO driver can insert the record, but it has other issues and missing functions.

**Comment by mklappen** [comment]
Thanks, appreciate it. I'm finding more and more that it's something other than the errors I thought in my previous posts. Definitely something with just the insert though... going through driver code trying to debug line by line... will keep at it for a little while and post back if I'm able to resolve. Unfortunately may have to stick with mysql for our setup for now.

**Comment by mklappen** [comment]
One final comment/question before I put this one to bed and move on to alternative solution...

unfortunately I'm still unable to determine why INSERT is not working for our setup. I keep coming back to a connected => false when I debug the connection from within cakephp's Table.php and Query.php. Can anyone clarify where this gets set exactly? 

In debugging the driver I do not see any issues, I'm able to connect for select and update queries and actually the table sequence is getting called every time I unsuccessfully try to INSERT from the controller 'add' method so not certain why that parameter returns "false"... at this time it's all I can come up with. Did side by side debug w/ mysql (which works fine) and that's the only apparent difference I could find. 

Table.php - from [protected] _statement => object(Cake\Oracle\Statement\OracleStatement) {
<code>[protected] _driver => object(Cake\Oracle\Driver\Oracle) {
    'connected' => false
}</code>

Query.php
    <code>object(Cake\Database\Connection) {
    'config' => [
        'password' => '****_',
        'username' => '**_**',
        'database' => '*****',
        'driver' => 'Cake\Oracle\Driver\Oracle',
        'name' => 'default'
    ],
    'driver' => object(Cake\Oracle\Driver\Oracle) {
        'connected' => false  
    },</code>

Will try to come back to it as some point in future. thanks.

**Comment by kasmonaut** [comment]
Hi, the only clue I have is that record is being inserted just fine, but
then CakePHP is trying to do more,probably get ID of new record,FAILS on
that and roles back the whole transaction. After trying few things on my
box, that is the only decision i conclude. I think this OCI CakePHP driver
for Oracle needs to prepare new statement when trying to get id.

**Comment by snelg** [comment]
Sorry I couldn't join the conversation earlier.
As with everything else associated with this Cake driver, I've only coded the pieces I've needed for my own use so far. I've only been using the driver to _read_ some old legacy data, so I haven't used any inserts at all myself so far.
Based on what you've said, my hunch is that Cake is expecting the database to have some equivalent of MySQL's "last_insert_id". But Oracle didn't have any built-in "auto increment" concept at all before version 12c.
I'll see if I can recreate the problem myself, which will make it easier for me to see the exact problem.

**Comment by mklappen** [comment]
Thanks, I said I was going to stop working on this but still looking at it because it's bothering me :)

Appreciate everyone's replies.

What kasmonaut mentioned in latest post had me start looking in different direction. Basically the driver is returning a "connected => false". During the cakephp save process it goes through a number of methods/functions that establish connection and pass the query. 

I've finally traced where the query is not getting saved to the transactional(callable $callback) function in cakephp/cakephp/src/Database/Connection.php. The call to the transactional either rolls back or commits the result. The result is being return as false so it goes to the rollback option. 

I believe result is coming back as false due to the fact that the driver is incorrectly being set to "connected" => "false" even though it is connected... I tested this by changing the $this->rollback() call to $this->commit() in transactional function and it's inserting into DB w/o issue. 

If possible can you confirm where the Driver actually gets assigned the "connected" property? been scouring the code but haven't been able to trace it yet.  

thx

**Comment by snelg** [comment]
Oh good, I just precisely re-created your bug,

```
ORA-24374: define not done before fetch or execute and fetch [ROOT/vendor/yajra/laravel-pdo-via-oci8/src/yajra/Pdo/Oci8/Statement.php, line 231]
```

When I can re-create a bug, that means I can figure it out :)

**Comment by mklappen** [comment]
ok. I thought based on another issue string that ORA-24374 was not an issue as long as cakephp debug was turned off.

I guess I have morphed this issue string into something different with the connection=>false and being unable to insert... (that is, unless it is in fact due to the ORA-24374 error... )

**Comment by snelg** [comment]
Fixed!
Because PHP's "oci_num_rows" function returns 0 for SELECT statements, I wrote the driver to use Cake's "Buffered Results" by default. Otherwise any call to "numRows()" would not work properly.

But for INSERT, UPDATE, and DELETE statements, it's backwards: "oci_num_rows" _does_ return the number of affected rows, and Buffered Results always returns 0 (since no rows are actually returned from the query).

New release (1.0.6) is in packagist already, so "composer update" should get it.

**Comment by mklappen** [comment]
Awesome!! New code working for me, no ORA-24374 error and everything inserting w/o issue. On to implementing in my full app and converting to Oracle. 
Thank you, thank you, thank you... A week's worth of debugging and head pounding paid off! :)
